### PR TITLE
[Docs] Fix sphinx build warnings, improve C expressions

### LINKED ISSUE #4217: [BUG] 0.29.22: sphinx warnings
After add patch which fixes #4209 I found that sphinx emits some warnings.
```console
+ /usr/bin/python3 setup.py build_sphinx -b man --build-dir build/sphinx
Unable to find pgen, not compiling formal grammar.
running build_sphinx
Running Sphinx v4.0.2
making output directory... done
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: [new config] 53 added, 0 changed, 0 removed
reading sources... [100%] src/userguide/wrapping_CPlusPlus
CHANGES.rst:1694: WARNING: Inline emphasis start-string without end-string.
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/quickstart/install.rst:62: WARNING: duplicate citation Sage, other instance in /home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/quickstart/build.rst
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/related_work.rst:44: WARNING: duplicate citation Pyrex, other instance in /home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/quickstart/overview.rst
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/extension_types.rst:853: WARNING: Unexpected indentation.
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/memoryviews.rst:19: WARNING: Unknown target name: "pep 3118".
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/memoryviews.rst:230: WARNING: Unknown target name: "new style buffers".
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/memoryviews.rst:230: WARNING: Unknown target name: "pep 3118".
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/CONTRIBUTING.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/examples/README.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/directives.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/extension_types.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/index.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/interfacing_with_other_code.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/language_basics.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/limitations.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/reference/special_mention.rst: WARNING: document isn't included in any toctree
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/quickstart/overview.rst:46: WARNING: Citation [Cython] is not referenced.
done
writing... python-cython.3 { src/quickstart/index src/quickstart/overview src/quickstart/install src/quickstart/build src/quickstart/cythonize src/tutorial/index src/tutorial/cython_tutorial src/tutorial/external src/tutorial/clibraries src/tutorial/cdef_classes src/tutorial/pxd_files src/tutorial/caveats src/tutorial/profiling_tutorial src/tutorial/strings src/tutorial/memory_allocation src/tutorial/pure src/tutorial/numpy src/tutorial/array src/tutorial/readings src/tutorial/related_work src/tutorial/appendix src/userguide/index src/userguide/language_basics src/userguide/extension_types src/userguide/special_methods src/userguide/sharing_declarations src/userguide/external_C_code src/userguide/source_files_and_compilation src/userguide/early_binding_for_speed src/userguide/wrapping_CPlusPlus src/userguide/fusedtypes src/userguide/pypy src/userguide/limitations src/userguide/pyrex_differences src/userguide/memoryviews src/userguide/buffer src/userguide/parallelism src/userguide/debugging src/userguide/numpy_tutorial src/userguide/numpy_pythran src/changes } /home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/clibraries.rst:537: WARNING: Unparseable C cross-reference: 'void*'
Invalid C declaration: Expected identifier in nested name, got keyword: void [error at 4]
  void*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/clibraries.rst:567: WARNING: Unparseable C cross-reference: 'void*'
Invalid C declaration: Expected identifier in nested name, got keyword: void [error at 4]
  void*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/clibraries.rst:567: WARNING: Unparseable C cross-reference: 'void*'
Invalid C declaration: Expected identifier in nested name, got keyword: void [error at 4]
  void*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:164: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:169: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:323: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:347: WARNING: Unparseable C cross-reference: 'std::string'
Invalid C declaration: Expected end of definition. [error at 3]
  std::string
  ---^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:476: WARNING: Unparseable C cross-reference: 'char'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:490: WARNING: Unparseable C cross-reference: 'char'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:490: WARNING: Unparseable C cross-reference: 'char'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:508: WARNING: Unparseable C cross-reference: 'char'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:508: WARNING: Unparseable C cross-reference: 'unsigned char'
Invalid C declaration: Expected identifier in nested name, got keyword: unsigned [error at 8]
  unsigned char
  --------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:532: WARNING: Unparseable C cross-reference: 'long'
Invalid C declaration: Expected identifier in nested name, got keyword: long [error at 4]
  long
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:532: WARNING: Unparseable C cross-reference: 'unsigned long'
Invalid C declaration: Expected identifier in nested name, got keyword: unsigned [error at 8]
  unsigned long
  --------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:532: WARNING: Unparseable C cross-reference: 'int'
Invalid C declaration: Expected identifier in nested name, got keyword: int [error at 3]
  int
  ---^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:613: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:650: WARNING: Unparseable C cross-reference: 'wchar_t*'
Invalid C declaration: Expected end of definition. [error at 7]
  wchar_t*
  -------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:660: WARNING: Unparseable C cross-reference: 'Py_UNICODE*'
Invalid C declaration: Expected end of definition. [error at 10]
  Py_UNICODE*
  ----------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:660: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:666: WARNING: Unparseable C cross-reference: 'Py_UNICODE*'
Invalid C declaration: Expected end of definition. [error at 10]
  Py_UNICODE*
  ----------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/tutorial/strings.rst:686: WARNING: Unparseable C cross-reference: 'Py_UNICODE*'
Invalid C declaration: Expected end of definition. [error at 10]
  Py_UNICODE*
  ----------^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/language_basics.rst:245: WARNING: unknown keyword: None
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/language_basics.rst:245: WARNING: unknown keyword: False
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/language_basics.rst:245: WARNING: unknown keyword: NULL
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/special_methods.rst:14: WARNING: unknown keyword: cdef class
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/external_C_code.rst:98: WARNING: Unparseable C cross-reference: 'int'
Invalid C declaration: Expected identifier in nested name, got keyword: int [error at 3]
  int
  ---^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/external_C_code.rst:98: WARNING: Unparseable C cross-reference: 'int'
Invalid C declaration: Expected identifier in nested name, got keyword: int [error at 3]
  int
  ---^
/home/tkloczko/rpmbuild/BUILD/cython-0.29.22/docs/src/userguide/pypy.rst:66: WARNING: Unparseable C cross-reference: 'char*'
Invalid C declaration: Expected identifier in nested name, got keyword: char [error at 4]
  char*
  ----^
done
build succeeded, 45 warnings.
````

### PR DESCRIPTION
**Pull request by matusvalo**
Several warnings were fixed, C expressions were improved. There are still some warnings pending:

```
WARNING: unknown config value 'latex_paper_size' in override, ignoring
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 57 source files that are out of date
updating environment: [new config] 57 added, 0 changed, 0 removed
reading sources... [100%] src/userguide/wrapping_CPlusPlus
looking for now-outdated files... none found
pickling environment... done
checking consistency... /Users/matus/dev/cython/docs/src/reference/interfacing_with_other_code.rst: WARNING: document isn't included in any toctree
/Users/matus/dev/cython/docs/src/reference/limitations.rst: WARNING: document isn't included in any toctree
/Users/matus/dev/cython/docs/src/reference/special_mention.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [100%] src/userguide/wrapping_CPlusPlus
WARNING: LaTeX command 'latex' cannot be run (needed for math display), check the imgmath_latex setting
```

I  was not sure if .rst files not included in toctree can be removed or not (they are basically empty).

The rationale for migration `:c:type:` to `:c:expr:` is following link:

https://stackoverflow.com/questions/63080103/how-to-reference-built-in-c-types-with-rst

Fixes #4217

### CODE REVIEW

**Comment by da-woods** [comment]
> I was not sure if .rst files not included in toctree can be removed or not (they are basically empty).

I think src/reference is a mixture of
* files that where the documentation has been moved that redirect people (should be kept)
* genuinely useless empty pages (should be deleted)

I think sphinx `:orphan:` might hide the warning (https://www.sphinx-doc.org/en/master/usage/restructuredtext/field-lists.html#metadata)

**Comment by da-woods** [comment]
> genuinely useless empty pages (should be deleted)

Although might be better handled in a separate PR...

File: docs/src/userguide/language_basics.rst [hunk 1]
```diff
@@ -457,10 +457,10 @@ C functions, on the other hand, can have parameters of any type, since they're
 passed in directly using a normal C function call.
 
 C Functions declared using :keyword:`cdef` or the ``@cfunc`` decorator with a
-Python object return type, like Python functions, will return a :keyword:`None`
+Python object return type, like Python functions, will return a ``None`
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:460]:
```suggestion
Python object return type, like Python functions, will return a ``None``
```

**Review by da-woods — COMMENTED** [review]
This may be where I don't quite understand Sphinx but I'm not quite sure why we're using 
```
:c:expr:`void*`
```

instead of
```
``void*``
```

If I understand right, the values `:c:expr:` assumes that you might have a bunch of definitions of C functions or types, so it'd create a link to the definition of `void`. We don't have these definitions anywhere (or plan to create them) though so I'd think it's better to write them as "just code"

**Comment by matusvalo** [comment]
> If I understand right, the values :c:expr: assumes that you might have a bunch of definitions of C functions or types, so it'd create a link to the definition of void. We don't have these definitions anywhere (or plan to create them) though so I'd think it's better to write them as "just code"

Yes I understand it in the same way. The most trivial argument is the style of the result. If you use `:c:expr:`, sphinx generates bold font (first one in example), if you use as:
```
``void*``
```
It generates non bold font (second one in example) - see the example having both styles:
image

The style currently in PR is consistent with the rest of C expressions. The argument against is that we are not consistent in the documentation (e.g. see the `int` type in example). I do not have hard preference, so @da-woods let me know if you still want to change it as you propose and I will do it.

**Comment by matusvalo** [comment]
> I think sphinx :orphan: might hide the warning (https://www.sphinx-doc.org/en/master/usage/restructuredtext/field-lists.html#metadata)

I will create separate PR for removing these files so I am not going to add `:orphan:` to them.

**Comment by da-woods** [comment]
> I do not have hard preference, so @da-woods let me know if you still want to change it as you propose and I will do it.

I don't think I do have a hard preference. Just trying to understand what the change is doing

**Comment by matusvalo** [comment]
My personal feeling is that we should stick with `:c:` statements for C expressions and `:py:` statements for python expressions.

File: docs/src/userguide/external_C_code.rst [hunk 2]
```diff
@@ -97,8 +97,8 @@ match the C ones, and in some cases they shouldn't or can't. In particular:
 
 #. If the header file uses macros to define constants, translate them into a
    normal external variable declaration.  You can also declare them as an
-   :keyword:`enum` if they contain normal :c:type:`int` values.  Note that
-   Cython considers :keyword:`enum` to be equivalent to :c:type:`int`, so do
+   :keyword:`enum` if they contain normal :c:expr:`int` values.  Note that
+   Cython considers :keyword:`enum` to be equivalent to :c:expr:`int`, so do
```
Comment (scoder) [inline docs/src/userguide/external_C_code.rst:101]: Shouldn't this use `:c:type:`?

File: docs/src/tutorial/strings.rst [hunk 3]
```diff
@@ -651,25 +651,25 @@ Windows and wide character APIs
 -------------------------------
 
 Windows system APIs natively support Unicode in the form of
-zero-terminated UTF-16 encoded :c:type:`wchar_t*` strings, so called
+zero-terminated UTF-16 encoded :c:expr:`wchar_t*` strings, so called
 "wide strings".
 
 By default, Windows builds of CPython define :c:type:`Py_UNICODE` as
-a synonym for :c:type:`wchar_t`. This makes internal :obj:`unicode`
+a synonym for :c:expr:`wchar_t`. This makes internal :obj:`unicode`
```
Comment (scoder) [inline docs/src/tutorial/strings.rst:658]: Can't we stick with `:c:type:` here?

**Review by scoder — COMMENTED** [review]
While this seems to aim for consistency, I'm not sure it's consistent in itself. What is the rule? Should it use `:c:type:` for any trivial type and `:c:expr:` for non-trivial types (like pointers)?

File: docs/src/userguide/external_C_code.rst (hunk 2 above)
Comment (matusvalo) [reply docs/src/userguide/external_C_code.rst:101]: When we use `:c:type:` sphinx is trying to find definition of `int` type which does not exist. This yields warnings during generating of documentation:
```
/Users/matus/dev/cython/docs/src/userguide/external_C_code.rst:98: WARNING: Unparseable C cross-reference: 'int'
Invalid C declaration: Expected identifier in nested name, got keyword: int [error at 3]
  int
  ---^
``` 
The fix is to use `c:expr:` instead - see https://stackoverflow.com/questions/63080103/how-to-reference-built-in-c-types-with-rst for details.

File: docs/src/tutorial/strings.rst (hunk 3 above)
Comment (matusvalo) [reply docs/src/tutorial/strings.rst:658]: There is no definition of `wchar_t` type. This is yielding errors during generating of documentation:
```
/Users/matus/dev/cython/docs/src/tutorial/strings.rst:653: WARNING: Unparseable C cross-reference: 'wchar_t*'
Invalid C declaration: Expected end of definition. [error at 7]
  wchar_t*
  -------^
```

File: docs/src/tutorial/clibraries.rst [hunk 4]
```diff
@@ -616,7 +616,7 @@ item value, so there is no way to explicitly signal an error to the
 calling code.
 
 The only way calling code can deal with this situation is to call
-``PyErr_Occurred()`` when returning from a function to check if an
+c:function:`PyErr_Occurred()` when returning from a function to check if an
```
Comment (da-woods) [inline docs/src/tutorial/clibraries.rst:619]:
```suggestion
:c:function:`PyErr_Occurred()` when returning from a function to check if an
```

**Comment by da-woods** [comment]
I'll merge this - thanks Matus.

I kind of think that these `:c:expr` tags may be more trouble than they're worth compared to the regular `` inline code blocks (just because they seem very fiddly and I'm not sure how useful the links back to the Python documentation are). But I believe they're being used correctly.

**Pull request merged**